### DISCUSSION TITLE
Which alternate installs for Apple M1 ?

### DISCUSSION THREAD

**Discussion by capwalker** [Homebrew/discussions#1587] [Getting started]
I got an Apple M1 Mac mini last month thinking it would be the best thing since sliced bread. Ah, no. There's some fine lines too small for me to read. Like you can't upgrade the memory. Buy what you might need first and expect the newer Mac Pro chip. And double check new new Homebrew syntax.

From all I can tell, there are numerous and different installs for Big Sur and M1. All that I've tried will use both /usr and /opt, I made sure to delete those subdirectories before I install. 

So I'd really need an authoritative "this is how to install".

**Comment by gromgit** [comment]
There aren't "numerous and different installs for Big Sur and M1", only two:

M1-based Homebrew (recommended):
```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

Intel-based Homebrew (if the software you need isn't M1-ready):
```
arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

**Comment by capwalker** [comment]
And how are we to carry over any prior setup ?

Adrian, on a private note: what are you trying to convey with "there aren't
..."  ? My impression was collaboration not aggression. It really wasn't so
important as to be a point of contention.

**Reply by gromgit** [reply]
> And how are we to carry over any prior setup ?

Assuming you're talking about carrying over your setup from an Intel Mac, the simplest way:
```
# On the Intel Mac, dump your Homebrew install list to ${HOME}/.Brewfile
brew bundle dump --global --force

# Copy ${HOME}/.Brewfile to the same location on your M1 Mac, then...
brew bundle --global
```

> Adrian, on a private note: what are you trying to convey with "there aren't ..." ? My impression was collaboration not aggression. It really wasn't so important as to be a point of contention.

Sorry about that, I was reacting to what looked like pointless hyperbole to me.

And this was hardly a "private note". :grin:

**Comment by capwalker** [comment]
It's the M1  which I either erase the subdirectories and start over or
maybe reinstall the OS,  And this is also a question as to which method
would be best to start over.

**Reply by gromgit** [reply]
The same procedure I mentioned earlier applies; you just `brew bundle dump --global --force` on the M1 instead, after you've installed whatever you want. I suggest you try doing just that, then inspecting the contents of `${HOME}/.Brewfile` to understand just what's being dumped.

If you're just planning to wipe and reinstall Homebrew, there's no reason to reinstall the OS. Homebrew works very hard to not touch anything outside `/opt/homebrew` (or `/usr/local` for an Intel setup), so `rm -fr /opt/homebrew` should suffice.

**Comment by capwalker** [comment]
I'm not sure I'm phrasing my points correctly:

In installing on the M1 for the first time, I'm not sure of how I installed
Homebrew. I did notice that there were calls to /Usr and /Opt. I take full
responsibility there. That was cleared when I used the script you posted
(thanks).

I'm considering if I need to clean up/uninstall /usr. In that case I would
use the transfer method from usr to opt. Or I could opt to leave well
enough alone.

You are quite ahead of me in your scripts and homebrew knowledge. I'm just
now going beyond hardware, networking and Pascal.

**Comment by capwalker** [comment]
I spoke too soon, BREW DOCTOR complains about /usr/local files. I'm going
to try removing that directory AFTER a backup.

**Comment by sroby** [comment]
Homebrew uses the old /usr/local location for Intel and the new /opt/homebrew location for native ARM. They are separate installs that work independently from each other (and have their own brew executables). You should have a /usr/local installation only if you need software that is not available native, otherwise you can get rid of it.

**Comment by capwalker** [comment]
Thanks. Unfortunately in trying multiple installs they crossed. Sometimes
calling /opt sometimes /usr. Last night's install was almost perfect, Once
I cleaned up my end, it waas great.

I'm trying to get python running. I need to get past this :CFLAGS=-I$(brew
--prefix)/include LDFLAGS=-L$(brew --prefix)/lib pip install <package>
-bash: syntax error near unexpected token `newline'

It's definitely higher than my understanding, but it seems to fail at (brew
prefix)

Please anyone jump in...

**Reply by gromgit** [reply]
First off, I suggest to "view (and respond to)" any future responses in GitHub itself, instead of via email, because proper formatting of your questions (especially the commands/code portions) is **critical** to getting the right answers.

For example, because you responded via email, I'm seeing this:

> I'm trying to get python running. I need to get past this :CFLAGS=-I$(brew
> --prefix)/include LDFLAGS=-L$(brew --prefix)/lib pip install <package>
> -bash: syntax error near unexpected token `newline'

which I can only assume is really:

> I'm trying to get python running. I need to get past this :
> ```
> $ CFLAGS=-I$(brew
> --prefix)/include LDFLAGS=-L$(brew --prefix)/lib pip install <package>
> -bash: syntax error near unexpected token `newline'
> ```

If you literally have a line break in the command above, that's what `bash` (your command shell) is complaining about. A line break _ends_ a command, which means you're actually typing in _two separate commands_ above:

> ```
> CFLAGS=-I$(brew
> ```
and:
> ```
> --prefix)/include LDFLAGS=-L$(brew --prefix)/lib pip install <package>
> ```
Note the unmatched parenthesis in each command.

What you really want a single command, like this:
```
CFLAGS=-I$(brew --prefix)/include LDFLAGS=-L$(brew --prefix)/lib pip install <package>
```

If that's not the problem you're facing, then you can blame your email client for reformatting your message...and post followups on this page instead, with proper formatting.

**Comment by capwalker** [comment]
This goes back to my 'hyperbole' (LOL) since I searched for an install
everywhere not just the official versions, I really had screwed up the
initial install. From all I can tell now, it's straight. Some old artiles
using the old syntax I have to translate, but I have it working. Thanks.
